### PR TITLE
[3.4] etcdserver: add cluster id check for hashKVHandler

### PR DESCRIPTION
**Pull request by ahrtr**
backport https://github.com/etcd-io/etcd/pull/15924 to 3.4

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: etcdserver/corrupt_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	var revision = 1
+
+	etcdSrv := &EtcdServer{}
+	etcdSrv.cluster = newTestCluster(nil)
+	etcdSrv.cluster.SetID(types.ID(localClusterID), types.ID(localClusterID))
+	be, _ := betesting.NewDefaultTmpBackend()
+	defer func() {
+		assert.NoError(t, be.Close())
+	}()
+	//etcdSrv.kv = mvcc.New(zap.NewNop(), be, &lease.FakeLessor{}, mvcc.StoreConfig{})
```
Comment (serathius) [inline etcdserver/corrupt_test.go:49]:
```suggestion
```

File: etcdserver/corrupt_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+				}
+				return
+			}
+
+			hashKVResponse := pb.HashKVResponse{}
+			err = json.Unmarshal(body, &hashKVResponse)
+			if err != nil {
+				t.Fatalf("unmarshal response error: %v", err)
+			}
+			//hashValue, _, err := etcdSrv.KV().HashStorage().HashByRev(int64(revision))
```
Comment (serathius) [inline etcdserver/corrupt_test.go:114]:
```suggestion
```

File: etcdserver/corrupt_test.go (hunk 1 above)
Comment (ahrtr) [reply etcdserver/corrupt_test.go]: done & thx

File: etcdserver/corrupt_test.go (hunk 2 above)
Comment (ahrtr) [reply etcdserver/corrupt_test.go]: done & thx

**Comment by serathius** [comment]
For posterity it's worth to mention this backport also includes improvements to `TestInPlaceRecovery` from https://github.com/etcd-io/etcd/pull/16677.

**Review by serathius — APPROVED** [review]

**Pull request merged**